psl-core/src/main/java/org/linqs/psl/util/ModelDataCollector.java

```diff
@@ -0,0 +1,332 @@
+package org.linqs.psl.util;
+
+import org.linqs.psl.grounding.GroundRuleStore;
+import org.linqs.psl.model.atom.Atom;
+import org.linqs.psl.model.atom.GroundAtom;
+import org.linqs.psl.model.atom.QueryAtom;
+import org.linqs.psl.model.formula.AbstractBranchFormula;
+import org.linqs.psl.model.formula.Conjunction;
+import org.linqs.psl.model.formula.Disjunction;
+import org.linqs.psl.model.formula.Formula;
+import org.linqs.psl.model.formula.Implication;
+import org.linqs.psl.model.formula.Negation;
+import org.linqs.psl.model.rule.AbstractRule;
+import org.linqs.psl.model.rule.arithmetic.AbstractGroundArithmeticRule;
+import org.linqs.psl.model.rule.GroundRule;
+import org.linqs.psl.model.rule.logical.AbstractLogicalRule;
+import org.linqs.psl.model.rule.Rule;
+import org.linqs.psl.model.rule.UnweightedGroundRule;
+import org.linqs.psl.model.rule.WeightedGroundRule;
+import org.linqs.psl.model.rule.WeightedRule;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.model.term.Term;
+import org.linqs.psl.model.term.Variable;
+
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+public class ModelDataCollector {
+    private static final Logger log = LoggerFactory.getLogger(ModelDataCollector.class);
+
+    private static Runtime runtime = null;
+    private static ModelData modelData = null;
+
+    private static String outputPath = null;
+
+    static {
+        init();
+    }
+
+    private ModelDataCollector() {}
+
+    private static synchronized void init() {
+        if (runtime != null) {
+            return;
+        }
+        modelData = new ModelData();
+        runtime = Runtime.getRuntime();
+        runtime.addShutdownHook(new ShutdownHook());
+    }
+
+    public static void outputJSON() {
+        FilterOutputStream stream = System.out;
+
+        try {
+            if (outputPath != null) {
+                stream = new GZIPOutputStream(new PrintStream(outputPath));
+            }
+
+            writeToStream(stream);
+
+            if (outputPath != null) {
+                stream.close();
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException("Could not write to path: " + outputPath, ex);
+        }
+    }
+
+    /**
+     * Write to stream with JSON formatting.
+     */
+    private static void writeToStream(FilterOutputStream stream) throws IOException {
+        stream.write("{\"truthMap\":".getBytes());
+
+        // Write each map as a JSON object, each JSON object is comma delimited.
```

Comment:
Move this comment to the top of this method (first line inside this method).

Then, group these lines so it is more clear what goes together.

For example:
```
        // Write each map as a JSON object, each JSON object is comma delimited.
        stream.write("{");

        stream.write("\"truthMap\":".getBytes());
        writeMap(stream, vizData.truthMap, "truthMap");
        stream.write(",");

        stream.write("\"rules\":".getBytes());
        writeMap(stream, vizData.rules, "rules");
        stream.write(",");

        stream.write("\"groundRules\":".getBytes());
        writeMap(stream, vizData.groundRules, "groundRules");
        stream.write(",");

        stream.write("\"groundAtoms\":".getBytes());
        writeMap(stream, vizData.groundAtoms, "groundAtoms");

        stream.write("}");
```

This uses more lines, but makes it more clear to the reader how things are done.